### PR TITLE
testing tools + mocks

### PR DESCRIPTION
**Pull request by nl0**
* `feature` interface:
   * add step back support
   * add support for running code in `afterEach` block
* mocks for some components
* `fetch-mocks` for request testing
* various reusable step definitions
* `testing/util`: refactor, new helpers

### CODE REVIEW

**Comment by nl0** [comment]
@akarve rtm

File: catalog/app/testing/util.js [hunk 1]
```diff
@@ -1,17 +1,143 @@
...
+
+  Object.assign(Mock.prototype, methods);
+
+  return Mock;
+};
+
+/**
+ * Create a selector (suitable for enzyme or cheerio wrappers) that selects
+ * a mock component created by `mockComponent[Class]` or its nested element / prop.
+ *
```
Comment (akarve) [inline catalog/app/testing/util.js]: do we have to have an empty line between `@param` calls? it's not a big deal but we should be consistent across the project.

File: catalog/app/testing/RequestTracker.js [hunk 2]
```diff
@@ -1,70 +0,0 @@
-import invariant from 'invariant';
```
Comment (akarve) [inline catalog/app/testing/RequestTracker.js]: What became of the logic in this file?

**Comment by nl0** [comment]
@akarve 
> looks reasonable but (maybe this happened in a separate PR?) need docs on how tools works and basics of writing new tests.

yep, it's in my backlog, I'll write it up soon

File: catalog/app/testing/RequestTracker.js (hunk 2 above)
Comment (nl0) [reply catalog/app/testing/RequestTracker.js]: `fetch-mock` package provides similar functionality

File: catalog/app/testing/util.js (hunk 1 above)
Comment (nl0) [reply catalog/app/testing/util.js]: I usually put an empty line if params are not related and/or they have multiline description. If params are related (example: object props) and have short descriptions, I usually don't put an empty line.

File: catalog/app/testing/feature.js [hunk 3]
```diff
@@ -99,15 +138,23 @@ const renderSteps = (steps, getCtx, next) => {
...
+    renderSteps(rest, getCtx, next);
+    return;
+  }
+
   if (steps[0].type === Then) {
-    const isNotWhen = ({ type }) => type !== When;
-    const thens = takeWhile(steps, isNotWhen);
-    const rest = dropWhile(steps, isNotWhen);
+    const isThen = ({ type }) => type !== When && type !== Branch;
+    const thens = takeWhile(steps, isThen);
```
Comment (meffij) [inline catalog/app/testing/feature.js]: [partition](https://lodash.com/docs/4.17.10#partition) is a function that expresses your intent here better I think

**Comment by nl0** [comment]
> I agree with Aneesh that I'd like to see some documentation of how this testing strategy works, maybe some rationale given how heavy it is.

@akarve @meffij I wanted to do it in a subsequent PR if you don't mind

File: catalog/app/testing/feature.js (hunk 3 above)
Comment (nl0) [reply catalog/app/testing/feature.js]: @meffij `partition` does something slightly different: it puts all the elements satisfying the predicate into the first group, and all the others into the second, but i need to split the array at the first occasion of the element satisfying the predicate

**Comment by akarve** [comment]
@nl0 OK to doc in separate PR but from this point forward all PRs should contain their own docs. makes review easier. too easy to forget docs otherwise.

**Comment by nl0** [comment]
@akarve sure, it totally makes sense

File: catalog/app/testing/feature.js (hunk 3 above)
Comment (meffij) [reply catalog/app/testing/feature.js]: I see, so you aren't guaranteed to take the whole array.

**Comment by meffij** [comment]
For reference, this PR is being merged because it's low-risk, but more effort is needed to make sure we understand the new testing ideas and that we're sure everything works correctly.

**Pull request merged**